teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,605 @@
+  for code injection and, by contrast, difficult for a `Task` _user_ to verify that
+  such an injection can't or hasn't taken place.
+- **Reliability**: It is easy for a `Task` _user_ to accidentally pass in a `Parameter`
+  with a character that would make the `Script` invalid and fail the `Task`, making
+  the `Task` extremely fragile.
+
+  The aim of this TEP is to introduce a safer and more reliable method to access `Parameter` variables from `Scripts`.
+
+## Motivation
+*Tekton Pipelines* provides variables used to inject values into the contents of certain
+fields, including `Scripts`. The values you can inject come from a range of sources,
+including `Parameters`. Interpolating `Parameters` in `Script` presents two main challenges:
+**Security** and **Reliability**. In this section, we provide the background information and
+expound on the challenges users face when interpolating `Parameters` in `Scripts`.
+
+As [documented][params-vars-docs], this is the syntax of `Parameter` interpolation,
+where `<name>` is the name of the `Parameter`:
+
+```yaml
+# dot notation
+$(params.<name>)
+# or bracket notation (wrapping <name> with either single or double quotes):
+$(params['<name>'])
+$(params["<name>"])
+``` 
+
+The mechanism of variable substitution is a string replacement that's performed by the
+*Tekton Controller* when a `TaskRun` is executed.
+
+Common examples of user-generated input that could be passed via `Parameter` and
+interpolated in a `Script` include:
+* Commit messages
+* Pull request descriptions and comments
+* Image names or registry URLs
+* Project or test suites names
+* Paths into `Workspaces`
+
+Tekton's own use of `Parameter` variables has evolved to avoid placing them
+directly in `Scripts`:
+1. Migrated our plumbing repository from using `Parameters` directly
+   (e.g.`$(params.foo)`) in `Script` to passing them in through environment variables
+   (e.g. `${FOO}`).  This was a precaution against possible code injections to our
+   continuous integration:
+     * [plumbing-pr-973]
+     * [plumbing-pr-974]
+     * [plumbing-pr-975]
+     * [plumbing-pr-976]
+     * [plumbing-pr-977]
+2. [Migrated][catalog-pr-711] the `git-clone` Catalog `Task` to use environment variables
+   instead of direct `Parameter` injection and, in doing so, made a backwards incompatible
+   change to it.
+3. Host [guidance in the *Tekton Catalog*][catalog-recommendation] warning against
+   `Parameters` in `Scripts`.
+
+Given that we have recognized a flaw in our own usage, it bears exploring whether
+this means there are flaws in the tool itself that allow this to occur. The flaws
+were previously discussed in [TEP-0017][tep-0017], but we pushed back deciding on
+a solution in favor of documenting the existing workarounds. It is time to revisit
+the problem and explore a solution considering our:
+1. Move towards V1 - decide on path forward ahead of stabilizing the API
```

Comment:
the `v1` API is out already - does this refer to Tekton Pipelines v1.x ?

Reply:
Updated the text there now - yes exactly, referring to v1.x or a v2 API.